### ISSUE TITLE
Additional JsonNode functionality

### ISSUE THREAD

**Issue by steveharter** [dotnet/runtime#56592]
### Background and motivation

Add missing features that were requested by the community. Note that `System.Text.Json.Nodes.JsonNode` and derived types are new, added in 6.0 preview 4.

### API Proposal

```diff
namespace System.Text.Json.Nodes
{
     public abstract class JsonNode
     {
         // In read mode, simply clones the underlying JsonElement without any possible JSON comments.
         // In edit mode this is expensive for values wrapping a non-node value since it need to serialize to JsonNode.
+        public JsonNode DeepClone();

         // In read mode, simply compares against the underlying JsonElement.
         // In edit mode this is expensive for values wrapping a non-node value since it need to serialize to JsonNode.
         // Made `static` for easier `null` comparison
+        public static bool DeepEquals(JsonNode? node1, JsonNode? node2);

         // In read mode, simply uses JsonElement.ValueKind.
         // In edit mode for values wrapping a non-node value that is not a known type (string, int, long, etc) then
         // JsonValueKind.Undefined is returned implying that to return the actual ValueKind one would serialize to a
         // JsonElement\JsonNode along with any possible custom converter.
+        public JsonValueKind GetValueKind()

         // Returns property name from parent object. Parent must be JsonObject otherwise InvalidOperationException.
+        public string GetPropertyName();

         // Returns index from parent array. Parent must be JsonArray otherwise InvalidOperationException.
+        public int GetElementIndex();

         // Replace this instance with a new value. The parent object\array is updated accordingly.
+        void ReplaceWith<T>(T value);
    }

    public class JsonArray
    {
         // For homogeneous arrays, returns enumerator that wraps calls to GetValue<T>().
+        IEnumerable<T> GetValues<T>();
    }
}
```     

### API Usage
DeepClone and DeepEquals
```C#
JsonNode node = JsonNode.Parse("{\"Prop\":{\"NestedProp\":42}}");
JsonNode other = node.DeepClone();
bool same = JsonNode.DeepEquals(node, other); // true
```

GetValueKind
```cs
JsonNode node = JsonNode.Parse("42");
// In read-mode this is efficient:
JsonValueKind kind = node.ValueKind; // JsonValueKind.Number
// In edit-mode , Undefined is returned
node = new JsonValue(43);
kind = node.ValueKind; // JsonValueKind.Undefined; there could be a custom converter for in32 writing it as a JSON object
// Slower workaround:
kind = node.Deserialize<JsonElement>().ValueKind; // JsonValueKind.Number

node = JsonValue.Create(new Dictionary<string, int>());
kind = node.ValueKind; // JsonValueKind.Undefined 
// Slower workaround; note that dictionaries are written as JSON objects by default
kind = node.Deserialize<JsonElement>().ValueKind; // JsonValueKind.Object
```

JsonArray.GetValues
```cs
JsonArray jsonArray = new JsonArray(1, 2, 3, 2);

// To get all "2" values out today using LINQ requires something like:
IEnumerable<int> values = jsonArray.Select(x => x.GetValue<int>()).Where(i => i == 2);

// With the proposed GetValues this is simpler:
IEnumerable<int> values = jsonArray.GetValues<int>().Where(i => i == 2);
```

**Comment by campersau** [comment]
`GetPropertyIndex()` sounds wired to me what about `GetArrayIndex()`?

**Comment by eiriktsarpalis** [comment]
> GetPropertyIndex() sounds wired to me what about GetArrayIndex()?

I'd probably call this `GetElementIndex()`.

@steveharter not sure I understand how `GetAccessorNameFromPath()` is different than the existing `GetPath()` method.

**Comment by steveharter** [comment]
> I'd probably call this GetElementIndex().

Yep thanks.

> @steveharter not sure I understand how GetAccessorNameFromPath() is different than the existing GetPath() method.

It returns just the immediate parent's property name \ index. So given "$.Order.Customer.Address", calling that method on Address instance returns ".Customer"

**Comment by steveharter** [comment]
There are basically two scenarios for the accessor name as I see it:
- Obtaining the string property name ("Address") or integer index value
- Obtaining either the property name or integer value in string JsonPath syntax ("$.Address" or "[1]") for use with JsonPath. Since we don't have any JsonPath functionality (like querying) today other than the Path I don't think this that relevant.

**Comment by steveharter** [comment]
These APIs are not blocking as there are workarounds as shown in https://github.com/dotnet/runtime/issues/55827, so they don't meet the bar for v6 this late in the release.

**Comment by PascalSenn** [comment]
@steveharter I am currently rewriting JsonNodes to a different structure. It would be great if there was a Replace on JsonArray and JsonObject that does not require to remove the parent first.
Did you consider something like this?

**Comment by steveharter** [comment]
>  It would be great if there was a Replace on JsonArray and JsonObject that does not require to remove the parent first.
Did you consider something like this?

Thanks. I added Replace() semantics to JsonArray\JsonObject to the proposal.

The reason why `Replace()` wasn't added is there is a simple work-around of remove+add. However, does that does require the index or property name to be specified, so I can see why `Replace()` could be useful.

**Comment by slang25** [comment]
I'm here to upvote the `DeepClone` functionality, could really do with that 🙂

**Comment by terrajobst** [comment]
* `XDocument` uses a copy constructor on `XElement` instead, but `DeepClone` seems more discoverable.
* `IsRoot` doesn't seem that value, as the canonical way would be to check `node.Parent is null`.
* `GetValueKind()` should try to return a better value than `Undefined` such that modifications of primitives doesn't result in unexpected changes from, say, `Number` to `Undefined`.
    - Do we need the options? We might need it for settings such as the one that serializes numbers as strings
* In `XDocument` the `Replace` method is called `ReplaceWith`, which makes it a bit more clear that it's in-place semantics
* We need to define what `JsonArray.GetValues<object>` will do. Return the `JsonNode` seems the most viable, because boxing the underlying primitives wouldn't be well-defined.

```C#
namespace System.Text.Json.Nodes
{
     public partial class JsonNode
     {
        public JsonNode DeepClone();
        public static bool DeepEquals(JsonNode? node1, JsonNode? node2);
        public JsonValueKind GetValueKind(JsonSerializerOptions options = null);
        public string GetPropertyName();
        public int GetElementIndex();
        public void ReplaceWith<T>(T value);
    }
    public partial class JsonArray
    {
        public IEnumerable<T> GetValues<T>();
    }
}
```

**Comment by steveharter** [comment]
Per the API review discussion, I looked at `JsonValueKind GetValueKind(options)` and determined that the `options` is not needed.

Basically, the algorithm would look like this:
```cs
public JsonValueKind GetValueKind()
{
    if (this is JsonElement element) return element.ValueKind;
    else if (TryGetKindFromKnownTypes(this, out kind)) return kind;
    else return JsonValueKind.Unknown;
}
```

I believe `options` was there previously because the thought is that there for non-`JsonElement` types the code would automatically do the expensive serialize(options) in order to obtain the kind. Now, we just return `JsonValueKind.Unknown`.

**Comment by sebastienros** [comment]
```c#
node = new JsonValue(43);
kind = node.ValueKind; // JsonValueKind.Undefined; there could be a custom converter for in32 writing it as a JSON object
```

Doesn't seem to match the description:

```c#
         // In read mode, simply uses JsonElement.ValueKind.
         // In edit mode for values wrapping a non-node value that is not a known type (string, int, long, etc) then
         // JsonValueKind.Undefined is returned implying that to return the actual ValueKind one would serialize to a
         // JsonElement\JsonNode along with any possible custom converter.
         public JsonValueKind GetValueKind()
```

Or is `Int32` not a "known type" and I misunderstood the api doc?

That would be very unfortunate IMO. I understand that it could be something else when serialized, but at this point it's not serialized, and as a user I would definitely like it to return `JsonValueKind.Number`.

**Comment by eiriktsarpalis** [comment]
> Doesn't seem to match the description:

Based on @steveharter's latest comment I believe that description is out of date. The current iteration would always return `Undefined` for all wrapped .NET values.

> Per the API review discussion, I looked at JsonValueKind GetValueKind(options) and determined that the options is not needed.

In that case, couldn't it just be a property?

**Comment by davidhmays** [comment]
Any status on adding json merge functionality?
I keep going in circles seeing where merging JsonDocuments on a key has been requested, merged into other requests, than apparently lost.

Was really hoping to have a simple way to merge on a key, and am frustrated the "upgrade" doesn't offer this.

**Comment by goldsam** [comment]
@eiriktsarpalis You closed #31433 in favor of this issue, but merge functionality doesn't seem to have been addressed. Please reopen #31433 so ensure this critical functionality is not neglected.
